smart_contract/executor/src/default/tokens.rs

```diff
@@ -29,27 +29,64 @@ use crate::permission::{self, Token as _};
 /// ```
 macro_rules! declare_tokens {
     ($($($token_path:ident ::)+ { $token_ty:ident }),+ $(,)?) => {
-        macro_rules! map_token {
-            ($token:ident => $callback:ident) => {
-                match $token.definition_id().as_ref() { $(
+        macro_rules! map_token_type {
+            ($callback:ident) => { $(
+                $callback!($($token_path::)+$token_ty); )+
+            };
+        }
+
+        /// Enum with every default token
+        #[allow(clippy::enum_variant_names)]
+        #[derive(Clone)]
+        pub(crate) enum AnyPermissionToken {
+            $(
+                $token_ty($($token_path::)+$token_ty),
+            )*
+        }
+
+        impl TryFrom<::iroha_smart_contract::data_model::permission::PermissionToken> for AnyPermissionToken {
+            type Error = $crate::permission::PermissionTokenConversionError;
+
+            fn try_from(token: ::iroha_smart_contract::data_model::permission::PermissionToken) -> Result<Self, Self::Error> {
+                match token.definition_id().as_ref() { $(
                     stringify!($token_ty) => {
-                        if let Ok(token) = <$($token_path::)+$token_ty>::try_from($token.clone()) {
-                            $callback!(token);
-                        }
+                        let token = <$($token_path::)+$token_ty>::try_from(token)?;
```

Comment:
I don't like that we have `TryFrom<PermissionToken>` implemented for every token. The reason is that such construction is suboptimal. I would prefer removing it and inline the code here, as it's the only place where it's used

Reply:
@mversic what do you mean by suboptimal?

Inlining code here is not that simple since we generating `TryFrom` conversion for concrete permission inside procmacro. 

> it's the only place where it's used
is it possible that we will expect certain permission type? so that we can directly parse into specific type omitting `AnyPermissionToken`

Reply:
> @mversic what do you mean by suboptimal?

how it's been used mostly was that a programmer would just try guessing to which type it can be converted when the match would have done the job in one jump

> Inlining code here is not that simple since we generating TryFrom conversion for concrete permission inside procmacro.

I believe you can literally just copy paste that code

Reply:
Currently `TryFrom` is used to implement `Token` trait (in `is_owned_by` method) where we convert `PermissionToken` to the concrete type.  

And `is_owned_by` use with concrete tokens, not only `AnyPermissionToken`.

I would suggest to extract this refactor into separate task.